dataengineeringutils3/writer.py

```diff
@@ -1,42 +1,246 @@
 import sys
+import os
+import gzip
+import boto3
 
-from dataengineeringutils3.s3 import gzip_string_write_to_s3
+from dataengineeringutils3.s3 import gzip_string_write_to_s3, s3_path_to_bucket_key
 
+from io import BytesIO, StringIO
 
-class SplitFileWriter:
+
+class BaseSplitFileWriter:
     """
-    Base class for splitting large datasets in to chunks and writing to s3
+    Base class for splitting large datasets in to chunks and writing to s3.
+    This is acts as a file like object. Data is written to an in memory file
+    (note this file is defined in subclasses and not set in this base class).
+    until it hits a max_bytes limit at which point the data is written to S3
+    as single file. The in memory file is defined by the sub classes 
+    BytesSlitFileWriter and StringSplitFileWriter. These subclasses attempt
+    to mimic the expected response of BytesIO and StringIO. 
 
-    The extension and the _write methods are defined in classes which extend this class
+    :param s3_basepath: The base path to the s3 location you want to write to S3://...
+    :param filename_prefix: The filename that you want to keep constant. Every written file is prefixed with this string.
+    S3 objects written will end in the file number and the extension.
+    :param max_bytes: The maximum number of bytes for each file (uncompressed file size) default set at 1GB.
+    :param compress_on_upload: If the file should be compressed before writing to S3 (default True). Note does not affect 
+    the file_extension parameter.
+    :param file_extension: String representing the file extension. Should not be prefixed with a '.'.
+
+    :Example:
+
+    from dataengineeringutils3.writer import BytesSplitFileWriter
```

Comment:
The code looks fine but I think this example could do with more work.  

- It's a bit confusing that the 'test dict' is used repeatedly.  Need at least a comment to say you're using it repeadly to simulate a multi line file
- Rather than `# Same functionality passing ` would be good to have a comment that says what this really does.  Which is to say, every time you hit max bytes, will write a new file in s3.